### PR TITLE
Guarantee thread-safe for SDImageCachesManager SDImageCodersManager SDImageLoadersManager

### PR DESCRIPTION
**Pull request by zhongwuzw**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [ ] I have added the required tests to prove the fix/feature I am adding
* [ ] I have updated the documentation (if necessary)
* [ ] I have run the tests and they pass
* [ ] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: ...

### Pull Request Description

Thread safe of `caches`.

### CODE REVIEW

**Comment by dreampiggy** [comment]
This just ensure the outer call of `self.caches` property ( actually, the caller outside of `SDImageCachesManager` ) using in the for-in loop thread-safe.

However, this case is really rare. So this is why all `SDImageCachesManager`, `SDImageCodersManager`, `SDImageLoadersManager` use the same code.

If you want to change, don't change just one class. Change both of them together...

**Comment by zhongwuzw** [comment]
@dreampiggy ok, I'll change all of them. If we want to thread-safe, we need to ensure it's really thread-safe.

```
@property (nonatomic, copy, nullable) NSArray<id<SDImageCache>> *caches;
```

And I don't much like make `caches` read-write to public, I think we need make it only read, and only provide methods to manage caches, but if I change this, it's breaking, so let's keep it.

**Comment by dreampiggy** [comment]
@zhongwuzw This is actually, a setup logic in Appdelegate.m. So this is why it looks like this.

Consider this code:

```
// AppDelegate.m
- (void)applicationDidFinishLaunching:(UIApplication *)application {

// I want to use HEIF/BPG/SVG format
[SDImageCodersManager.sharedManager addCoder:SDImageHEIFCoder.sharedCoder];
[SDImageCodersManager.sharedManager addCoder:SDImageBPGCoder.sharedCoder];
[SDImageCodersManager.sharedManager addCoder:SDImageSVGCoder.sharedCoder];

// Or I can write like this?
SDImageCodersManager.sharedManager.coders = @[SDImageHEIFCoder.sharedCoder, SDImageBPGCoder.sharedCoder, SDImageSVGCoder.sharedCoder];

}
```

Which one is better...

**Comment by zhongwuzw** [comment]
@dreampiggy If I reimplement this, I will provide `addCoders:(NSArray*)`、`removeCoders:(NSArray*)`、`removeAllCoders`.

**Comment by dreampiggy** [comment]
We can just simply keeping only `add/remove` logic. But I think this is sometime, not so convenient for users. Especially, we have some default coders/loaders/caches here.

So if I want to supply only the target loaders I want. I have to for-in loop, and call remove on them....Not so good to look like this.

See code:

```objective-c
// I don't want the built-in `SDWebImageDownlaoder.sharedDownloader`, remove it!

for (id<SDImageLoader> *loader in SDImageLoadersManager.sharedManager.loaders) {
   [SDImageLoadersManager.sharedManager removeLoader:loader]
}
// add my own loader
[SDImageLoadersManager.sharedManager addLoader:MyOwnLoader.sharedLoader]
```

**Comment by dreampiggy** [comment]
> @dreampiggy If I reimplement this, I will only provide `addCoders:(NSArray*)`、`removeCoders:(NSArray*)`、`removeAllCoders`. And `caches` is read-only.

It's original written by @bpoplauschi 😂

I agree with this API actually...A better way than the read-write property.

File: SDWebImage/SDImageCachesManager.m [hunk 1]
```diff
@@ -17,6 +17,9 @@ @interface SDImageCachesManager ()
 @end
 
 @implementation SDImageCachesManager
+{
+    NSMutableArray<id<SDImageCache>> *caches_;
```
Comment (dreampiggy) [inline SDWebImage/SDImageCachesManager.m]: Emm...A little idea about coding style. Can we change this ivar naming to a more property one, like `mutableCaches` ? Placing a suffix `_` looks a little strange for most of Objective-C developers. It's more like a C-style

**Comment by dreampiggy** [comment]
Seems OK. After we change both of loaders/caches/coders magaer together, we can merge this into 5.0.0 to catch the last chance...

**Comment by zhongwuzw** [comment]
> > @dreampiggy If I reimplement this, I will only provide `addCoders:(NSArray*)`、`removeCoders:(NSArray*)`、`removeAllCoders`. And `caches` is read-only.
> 
> It's original written by @bpoplauschi 😂
> 
> I agree with this API actually...A better way than the read-write property.

But it's a API breaking, seems too late. 😄

**Comment by dreampiggy** [comment]
Seems CI is down becaues of codecov coverage service...I'll have a check...

**Pull request merged**